### PR TITLE
Fix #584 authorize transactions with non-immediate owner authorities

### LINKED ISSUE #584: Owner keys of non-immediately required accounts can not authorize a transaction
@pmconrad found the misbehavior [here](https://github.com/bitshares/bitshares-core/pull/570#issuecomment-357867908), quoted below:
> ..., if the general rule is that owner can always sign in place of active, then the recursion in sign_state::check_authority must also check for the owner auth (https://github.com/abitmore/bitshares-core/blob/97fe7208e62ef710ed16badd2404d738851965fd/libraries/chain/protocol/transaction.cpp#L199).

Say, if an account A's active authorities is
```
{
  "threshold": 1,
  "account_auths": [ "ID_of_account_B", 1 ],
  ...
}
```
then a transaction which requires account A's active authorities but only signed with account B's owner key should be accepted by the chain.

### PR DESCRIPTION
**Pull request by abitmore**
PR for #584.

Todo:
- [x] add test cases for APIs;
- [x] rebase after merged #1255;
  - [x] fix test cases added in #1255
<del>- [ ] update proposal tests after implemented #278.</del> update: moved this job to #278.
- [x] fix code added in #1384
<del>- [ ] fix test cases after #1360 is merged</del>
<del>- [ ] fix code after #1401 is merged</del>

### CODE REVIEW

**Comment by pmconrad** [comment]
Rebased on latest hardfork to resolve merge conflict

**Comment by abitmore** [comment]
@jmjatlanta, @oxarbitrage or @xeroc please review this so we can merge it?

**Comment by oxarbitrage** [comment]
i will want to take a look tomorrow after the feature release is done. thank you.

File: libraries/chain/include/graphene/chain/protocol/transaction.hpp [hunk 1]
```diff
@@ -213,9 +229,26 @@ namespace graphene { namespace chain {
...
 
+   /**
+    * Checks whether given public keys and approvals are sufficient to authorize given operations.
+    *   Throws an exception when failed.
+    *
+    * @param ops a vector of operations
+    * @param sigs a set of public keys
+    * @param get_active callback function to retrieve active authorities of a given account
+    * @param get_owner  callback function to retrieve owner authorities of a given account
+    * @param allow_non_immediate_owner whether to allow owner authrity of non-immediately
```
Comment (xeroc) [inline libraries/chain/include/graphene/chain/protocol/transaction.hpp]: typo `authrity`

**Pull request merged**